Deleting AS3 declaration with bigip-as3 resource still throws an errror

### Environment

 * TMOS/Bigip Version: BIG-IP 16.1.2.1 Build 0.0.10 Point Release 1 
 * Terraform Version: 1.2.8
 * Terraform bigip provider Version: 1.16.0

### Summary

Using big-as3 resource as part of the big-ip terraform provider works. But when an AS3 deployed app service gets deleted, it throws an error. This is the same issue as #620 which has has not been resolved, unfortunately.

### Steps To Reproduce

Steps to reproduce the behavior:

1. Terraform resource config to create the app service: 

(basically [Lab 2.1 Task 1](https://clouddocs.f5.com/training/community/big-iq-cloud-edition/html/class1/module2/lab1.html))

```
terraform {

    required_providers {
      bigip = {
          source = "F5Networks/bigip"
          version = "~>1.12"
      }
    }
}

provider "bigip" {
  address  = "host"
  port     = 8443
  username = "whatever"
  password = "secret"
}


locals {

  as3_json              = <<-EOT
    {
    "$schema": "https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json",
    "class": "AS3",
    "action": "deploy",
    "persist": true,
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.7.0",
        "id": "example-declaration-01",
        "label": "Task1",
        "remark": "Task 1 - HTTP Application Service",
        "Task1": {
            "class": "Tenant",
            "MyWebApp1http": {
                "class": "Application",
                "template": "http",
                "serviceMain": {
                    "class": "Service_HTTP",
                    "virtualAddresses": [
                        "10.38.1.138"
                    ],
                    "pool": "web_pool"
                },
                "web_pool": {
                    "class": "Pool",
                    "monitors": [
                        "http"
                    ],
                    "members": [
                        {
                            "servicePort": 80,
                            "serverAddresses": [
                                "192.168.100.1",
                                "192.168.100.2"
                            ],
                            "shareNodes": true
                        }
                    ]
                }
            }
        }
    }
}
    EOT
}

output "load_balancer_vips" {
    value = local.as3_json
    sensitive = false
}

resource "bigip_as3" "as3test" {

  as3_json      = local.as3_json
  tenant_filter = "Task1"
}
```
The above successfully creates the app service. Then we delete it by changing the declaration variable and execute the Terraform apply again (basically [Lab 2.4 Task 9](https://clouddocs.f5.com/training/community/big-iq-cloud-edition/html/class1/module2/lab4.html)):

```

locals {

  as3_json              = <<-EOT
    {
    "$schema": "https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json",
    "class": "AS3",
    "action": "deploy",
    "persist": true,
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.7.0",
        "id": "example-declaration-01",
        "label": "Task1",
        "remark": "Task 1 - HTTP Application Service",
        "Task1": {
            "class": "Tenant"
        }
    }
}
    EOT
}

```
3. Terraform debug log:

```
2022-12-02T17:45:41.037+0100 [INFO]  backend/local: apply calling Apply
2022-12-02T17:45:41.037+0100 [DEBUG] Building and walking apply graph for NormalMode plan
2022-12-02T17:45:41.037+0100 [DEBUG] ProviderTransformer: "bigip_as3.as3test" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/f5networks/bigip"]
2022-12-02T17:45:41.037+0100 [DEBUG] ProviderTransformer: "bigip_as3.as3test (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/f5networks/bigip"]
2022-12-02T17:45:41.037+0100 [DEBUG] ReferenceTransformer: "bigip_as3.as3test (expand)" references: []
2022-12-02T17:45:41.037+0100 [DEBUG] ReferenceTransformer: "local.as3_json (expand)" references: []
2022-12-02T17:45:41.037+0100 [DEBUG] ReferenceTransformer: "output.load_balancer_vips" references: [local.as3_json (expand)]
2022-12-02T17:45:41.038+0100 [DEBUG] ReferenceTransformer: "bigip_as3.as3test" references: [local.as3_json (expand)]
2022-12-02T17:45:41.038+0100 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/f5networks/bigip\"]" references: []
2022-12-02T17:45:41.038+0100 [DEBUG] Starting graph walk: walkApply
2022-12-02T17:45:41.038+0100 [DEBUG] created provider logger: level=debug
2022-12-02T17:45:41.039+0100 [INFO]  provider: configuring client automatic mTLS
2022-12-02T17:45:41.062+0100 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/f5networks/bigip/1.16.0/linux_amd64/terraform-provider-bigip_v1.16.0 args=[.terraform/providers/registry.terraform.io/f5networks/bigip/1.16.0/linux_amd64/terraform-provider-bigip_v1.16.0]
2022-12-02T17:45:41.063+0100 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/f5networks/bigip/1.16.0/linux_amd64/terraform-provider-bigip_v1.16.0 pid=35877
2022-12-02T17:45:41.063+0100 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/f5networks/bigip/1.16.0/linux_amd64/terraform-provider-bigip_v1.16.0
2022-12-02T17:45:41.091+0100 [INFO]  provider.terraform-provider-bigip_v1.16.0: configuring server automatic mTLS: timestamp=2022-12-02T17:45:41.090+0100
2022-12-02T17:45:41.131+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: plugin address: address=/tmp/plugin1795113059 network=unix timestamp=2022-12-02T17:45:41.131+0100
2022-12-02T17:45:41.131+0100 [DEBUG] provider: using plugin: version=5
2022-12-02T17:45:41.175+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2022-12-02T17:45:41.192+0100 [DEBUG] No provider meta schema returned
2022-12-02T17:45:41.243+0100 [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/f5networks/bigip\"]" changed the config value, but that value is unused
2022-12-02T17:45:41.244+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:41 [INFO] Initializing BigIP connection
2022-12-02T17:45:41.392+0100 [WARN]  Provider "registry.terraform.io/f5networks/bigip" produced an invalid plan for bigip_as3.as3test, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .as3_json: planned value cty.StringVal("{\"$schema\":\"https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json\",\"action\":\"deploy\",\"class\":\"AS3\",\"declaration\":{\"Task1\":{\"class\":\"Tenant\"},\"class\":\"ADC\",\"id\":\"example-declaration-01\",\"label\":\"Task1\",\"remark\":\"Task 1 - HTTP Application Service\",\"schemaVersion\":\"3.7.0\"},\"persist\":true}") does not match config value cty.StringVal("    {\n    \"$schema\": \"https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json\",\n    \"class\": \"AS3\",\n    \"action\": \"deploy\",\n    \"persist\": true,\n    \"declaration\": {\n        \"class\": \"ADC\",\n        \"schemaVersion\": \"3.7.0\",\n        \"id\": \"example-declaration-01\",\n        \"label\": \"Task1\",\n        \"remark\": \"Task 1 - HTTP Application Service\",\n        \"Task1\": {\n            \"class\": \"Tenant\"\n        }\n    }\n}\n") nor prior value cty.StringVal("{\"action\":\"deploy\",\"class\":\"AS3\",\"declaration\":{\"Task1\":{\"MyWebApp1http\":{\"class\":\"Application\",\"serviceMain\":{\"class\":\"Service_HTTP\",\"pool\":\"web_pool\",\"virtualAddresses\":[\"10.38.1.138\"]},\"template\":\"http\",\"web_pool\":{\"class\":\"Pool\",\"members\":[{\"serverAddresses\":[\"192.168.100.1\",\"192.168.100.2\"],\"servicePort\":80,\"shareNodes\":true}],\"monitors\":[\"http\"]}},\"class\":\"Tenant\"},\"class\":\"ADC\",\"id\":\"example-declaration-01\",\"label\":\"Task1\",\"remark\":\"Task 1 - HTTP Application Service\",\"schemaVersion\":\"3.7.0\"},\"persist\":true}")
      - .ignore_metadata: planned value cty.False for a non-computed attribute
bigip_as3.as3test: Modifying... [id=Task1]
2022-12-02T17:45:41.393+0100 [INFO]  Starting apply for bigip_as3.as3test
2022-12-02T17:45:41.393+0100 [DEBUG] bigip_as3.as3test: applying the planned Update change
2022-12-02T17:45:41.395+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:41 [INFO] Updating As3 Config :    {
2022-12-02T17:45:41.396+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     "$schema": "https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json",
2022-12-02T17:45:41.396+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     "class": "AS3",
2022-12-02T17:45:41.397+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     "action": "deploy",
2022-12-02T17:45:41.397+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     "persist": true,
2022-12-02T17:45:41.397+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     "declaration": {
2022-12-02T17:45:41.398+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "class": "ADC",
2022-12-02T17:45:41.398+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "schemaVersion": "3.7.0",
2022-12-02T17:45:41.398+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "id": "example-declaration-01",
2022-12-02T17:45:41.398+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "label": "Task1",
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "remark": "Task 1 - HTTP Application Service",
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         "Task1": {
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:             "class": "Tenant"
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:         }
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0:     }
2022-12-02T17:45:41.399+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: }
2022-12-02T17:45:41.775+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:41 [DEBUG] AS3 Version:3.40.0
2022-12-02T17:45:41.775+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:41 [DEBUG] Terraform Version:Terraform/1.2.8/terraform-provider-bigip/1.16.0
2022-12-02T17:45:42.760+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:42 [DEBUG]Code = 0,ID = 714ba3e7-4ddb-4ebb-9119-8568cdebd664
2022-12-02T17:45:50.240+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:50 [DEBUG]Sucessfully Created Application with ID  = 714ba3e7-4ddb-4ebb-9119-8568cdebd664
2022-12-02T17:45:50.240+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:50 [DEBUG] successfulTenants :Task1
2022-12-02T17:45:50.240+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:50 [INFO] Reading As3 config
2022-12-02T17:45:50.240+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:50 [DEBUG] Tenants in AS3 get call : Task1
2022-12-02T17:45:51.334+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:51 [DEBUG] AS3 json retreived from the GET call in Read function : 
2022-12-02T17:45:51.334+0100 [DEBUG] provider.terraform-provider-bigip_v1.16.0: 2022/12/02 17:45:51 [WARN] Json (Task1) not found, removing from state
2022-12-02T17:45:51.341+0100 [ERROR] vertex "bigip_as3.as3test" error: Provider produced inconsistent result after apply
2022-12-02T17:45:51.348+0100 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/f5networks/bigip/1.16.0/linux_amd64/terraform-provider-bigip_v1.16.0 pid=35877
2022-12-02T17:45:51.348+0100 [DEBUG] provider: plugin exited
```

### Expected Behavior
Resources are deleted and no error reported.

### Actual Behavior

Resources are deleted but an error is reported:

```
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to bigip_as3.as3test, provider "provider[\"registry.terraform.io/f5networks/bigip\"]" produced an unexpected new value: Root resource was present, but now
│ absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -265,7 +265,7 @@ func resourceBigipAs3Read(d *schema.ResourceData, meta interface{}) error {
 		if as3Resp == "" {
 			log.Printf("[WARN] Json (%s) not found, removing from state", d.Id())
 			_ = d.Set("as3_json", "")
-			//d.SetId("")
+			// d.SetId("")
 			return nil
 		}
 		_ = d.Set("as3_json", as3Resp)

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -50,7 +50,6 @@ func resourceBigipAs3() *schema.Resource {
 				Description: "AS3 json",
 				StateFunc: func(v interface{}) string {
 					json, _ := structure.NormalizeJsonString(v)
-
 					return json
 				},
 				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
@@ -265,7 +264,8 @@ func resourceBigipAs3Read(d *schema.ResourceData, meta interface{}) error {
 		}
 		if as3Resp == "" {
 			log.Printf("[WARN] Json (%s) not found, removing from state", d.Id())
-			d.SetId("")
+			_ = d.Set("as3_json", "")
+			//d.SetId("")
 			return nil
 		}
 		_ = d.Set("as3_json", as3Resp)

--- a/bigip/resource_bigip_do.go
+++ b/bigip/resource_bigip_do.go
@@ -41,9 +41,12 @@ func resourceBigipDo() *schema.Resource {
 				Required:    true,
 				Description: "DO json",
 				StateFunc: func(v interface{}) string {
-					json, _ := structure.NormalizeJsonString(v)
-					return json
+					jsonString, _ := structure.NormalizeJsonString(v)
+					return jsonString
 				},
+				//DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+				//	return true
+				//},
 			},
 			"timeout": {
 				Type:        schema.TypeInt,
@@ -316,12 +319,13 @@ func resourceBigipDoRead(d *schema.ResourceData, meta interface{}) error {
 	if resp.Status != "200 OK" {
 		return fmt.Errorf("Error while Sending/fetching http request :%s ", bodyString)
 	}
-
 	respRef1 := make(map[string]interface{})
 	if err := json.Unmarshal(respBody.Bytes(), &respRef1); err != nil {
 		return err
 	}
-	log.Printf("[DEBUG] in read resp_body is :%v", respRef1)
+	log.Printf("[INFO] in read resp_body is :%v", respRef1)
+	byteData, _ := json.Marshal(respRef1["declaration"])
+	_ = d.Set("do_json", string(byteData))
 
 	return nil
 

--- a/bigip/resource_bigip_do.go
+++ b/bigip/resource_bigip_do.go
@@ -44,9 +44,6 @@ func resourceBigipDo() *schema.Resource {
 					jsonString, _ := structure.NormalizeJsonString(v)
 					return jsonString
 				},
-				//DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
-				//	return true
-				//},
 			},
 			"timeout": {
 				Type:        schema.TypeInt,

--- a/bigip/resource_bigip_ltm_virtual_server.go
+++ b/bigip/resource_bigip_ltm_virtual_server.go
@@ -191,9 +191,9 @@ func resourceBigipLtmVirtualServer() *schema.Resource {
 				Computed: true,
 			},
 			"source_address_translation": {
-				Type:     schema.TypeString,
-				Optional: true,
-				//Computed:    true,
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
 				Description: "none, automap, snat",
 			},
 			"snatpool": {
@@ -388,9 +388,8 @@ func resourceBigipLtmVirtualServerRead(d *schema.ResourceData, meta interface{})
 	} else {
 		_ = d.Set("state", "disabled")
 	}
-	if _, ok := d.GetOk("source_address_translation"); ok {
-		_ = d.Set("source_address_translation", vs.SourceAddressTranslation.Type)
-	}
+	_ = d.Set("source_address_translation", vs.SourceAddressTranslation.Type)
+
 	_ = d.Set("snatpool", vs.SourceAddressTranslation.Pool)
 	_ = d.Set("policies", vs.Policies)
 	_ = d.Set("vlans", vs.Vlans)

--- a/bigip/resource_bigip_net_route_test.go
+++ b/bigip/resource_bigip_net_route_test.go
@@ -19,7 +19,7 @@ var TEST_ROUTE_NAME = fmt.Sprintf("/%s/test-route", TEST_PARTITION)
 var TEST_ROUTE_RESOURCE = `
 
 resource "bigip_net_vlan" "test-vlan" {
-	name = "` + TEST_VLAN_NAME + `"
+	name = "` + TestVlanName + `"
 	tag = 101
 	interfaces {
 		vlanport = 1.1
@@ -42,7 +42,7 @@ resource "bigip_net_route" "test-route" {
 var TEST_ROUTE_RESOURCE_UPDATE = `
 
 resource "bigip_net_vlan" "test-vlan" {
-        name = "` + TEST_VLAN_NAME + `"
+        name = "` + TestVlanName + `"
         tag = 101
         interfaces {
                 vlanport = 1.1

--- a/bigip/resource_bigip_net_selfip_test.go
+++ b/bigip/resource_bigip_net_selfip_test.go
@@ -21,7 +21,7 @@ var TEST_FLOAT_SELFIP_NAME = fmt.Sprintf("/%s/test-float-selfip", TEST_PARTITION
 
 var TEST_SELFIP_RESOURCE = `
 resource "bigip_net_vlan" "test-vlan" {
-  name = "` + TEST_VLAN_NAME + `"
+  name = "` + TestVlanName + `"
   tag = 101
   interfaces {
     vlanport = 1.1
@@ -58,10 +58,10 @@ func TestAccBigipNetselfip_create(t *testing.T) {
 					testCheckselfipExists(TEST_FLOAT_SELFIP_NAME),
 					resource.TestCheckResourceAttr("bigip_net_selfip.test-selfip", "name", TEST_SELFIP_NAME),
 					resource.TestCheckResourceAttr("bigip_net_selfip.test-selfip", "ip", "11.1.1.1/24"),
-					resource.TestCheckResourceAttr("bigip_net_selfip.test-selfip", "vlan", TEST_VLAN_NAME),
+					resource.TestCheckResourceAttr("bigip_net_selfip.test-selfip", "vlan", TestVlanName),
 					resource.TestCheckResourceAttr("bigip_net_selfip.test-float-selfip", "name", TEST_FLOAT_SELFIP_NAME),
 					resource.TestCheckResourceAttr("bigip_net_selfip.test-float-selfip", "ip", "11.1.1.2/24"),
-					resource.TestCheckResourceAttr("bigip_net_selfip.test-float-selfip", "vlan", TEST_VLAN_NAME),
+					resource.TestCheckResourceAttr("bigip_net_selfip.test-float-selfip", "vlan", TestVlanName),
 					resource.TestCheckResourceAttr("bigip_net_selfip.test-float-selfip", "traffic_group", "traffic-group-1"),
 				),
 			},
@@ -206,7 +206,7 @@ func TestAccBigipNetselfipRouteDomain(t *testing.T) {
 func testaccselfipRouteDomain(ip string) string {
 	resPrefix := `
 	resource "bigip_net_vlan" "test-vlan" {
-      name = "` + TEST_VLAN_NAME + `"
+      name = "` + TestVlanName + `"
 	  tag = 101
 	  interfaces {
 		vlanport = 1.1
@@ -226,7 +226,7 @@ func testaccselfipRouteDomain(ip string) string {
 func testaccselfipPortLockdownParam(portLockdown string) string {
 	resPrefix := `
 	resource "bigip_net_vlan" "test-vlan" {
-      name = "` + TEST_VLAN_NAME + `"
+      name = "` + TestVlanName + `"
 	  tag = 101
 	  interfaces {
 		vlanport = 1.1

--- a/bigip/resource_bigip_net_vlan.go
+++ b/bigip/resource_bigip_net_vlan.go
@@ -12,6 +12,7 @@ import (
 
 	bigip "github.com/f5devcentral/go-bigip"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
 )
 
 func resourceBigipNetVlan() *schema.Resource {
@@ -59,6 +60,13 @@ func resourceBigipNetVlan() *schema.Resource {
 					},
 				},
 			},
+			"cmp_hash": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.StringInSlice([]string{"default", "src-ip", "dst-ip"}, false),
+				Description:  "Specifies how the traffic on the VLAN will be disaggregated. The value selected determines the traffic disaggregation method",
+			},
 		},
 	}
 
@@ -74,13 +82,16 @@ func resourceBigipNetVlanCreate(d *schema.ResourceData, meta interface{}) error 
 
 	d.Partial(true)
 
-	err := client.CreateVlan(
-		name,
-		tag,
-	)
+	r := &bigip.Vlan{
+		Name:    name,
+		Tag:     tag,
+		CMPHash: d.Get("cmp_hash").(string),
+	}
+
+	err := client.CreateVlan(r)
 
 	if err != nil {
-		return fmt.Errorf("Error creating VLAN %s: %v", name, err)
+		return fmt.Errorf("Error creating VLAN %s: %v ", name, err)
 	}
 
 	d.SetId(name)
@@ -122,8 +133,9 @@ func resourceBigipNetVlanRead(d *schema.ResourceData, meta interface{}) error {
 		return nil
 	}
 
-	d.Set("name", vlan.FullPath)
-	d.Set("tag", vlan.Tag)
+	_ = d.Set("name", vlan.FullPath)
+	_ = d.Set("tag", vlan.Tag)
+	_ = d.Set("cmp_hash", vlan.CMPHash)
 
 	log.Printf("[DEBUG] Reading VLAN %s Interfaces", name)
 
@@ -165,13 +177,14 @@ func resourceBigipNetVlanUpdate(d *schema.ResourceData, meta interface{}) error 
 	log.Printf("[DEBUG] Updating VLAN %s", name)
 
 	r := &bigip.Vlan{
-		Name: name,
-		Tag:  d.Get("tag").(int),
+		Name:    name,
+		Tag:     d.Get("tag").(int),
+		CMPHash: d.Get("cmp_hash").(string),
 	}
 
 	err := client.ModifyVlan(name, r)
 	if err != nil {
-		return fmt.Errorf("Error modifying VLAN %s: %v", name, err)
+		return fmt.Errorf("Error modifying VLAN %s: %v ", name, err)
 	}
 
 	return resourceBigipNetVlanRead(d, meta)
@@ -186,9 +199,8 @@ func resourceBigipNetVlanDelete(d *schema.ResourceData, meta interface{}) error 
 
 	err := client.DeleteVlan(name)
 	if err != nil {
-		return fmt.Errorf("Error Deleting Vlan : %s", err)
+		return fmt.Errorf("Error Deleting Vlan : %s ", err)
 	}
-
 	d.SetId("")
 	return nil
 }

--- a/bigip/resource_bigip_net_vlan_test.go
+++ b/bigip/resource_bigip_net_vlan_test.go
@@ -16,9 +16,9 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/terraform"
 )
 
-var TEST_VLAN_NAME = fmt.Sprintf("/%s/test-vlan", TEST_PARTITION)
+var TestVlanName = fmt.Sprintf("/%s/test-vlan", TEST_PARTITION)
 
-var TEST_VLAN_RESOURCE = `
+var TestVlanResource = `
 resource "bigip_net_vlan" "test-vlan" {
 	name = "/Common/test-vlan"
 	tag = 101
@@ -38,9 +38,9 @@ func TestAccBigipNetvlan_create(t *testing.T) {
 		CheckDestroy: testCheckvlansDestroyed,
 		Steps: []resource.TestStep{
 			{
-				Config: TEST_VLAN_RESOURCE,
+				Config: TestVlanResource,
 				Check: resource.ComposeTestCheckFunc(
-					testCheckvlanExists(TEST_VLAN_NAME, true),
+					testCheckvlanExists(TestVlanName, true),
 					resource.TestCheckResourceAttr("bigip_net_vlan.test-vlan", "name", "/Common/test-vlan"),
 					resource.TestCheckResourceAttr("bigip_net_vlan.test-vlan", "tag", "101"),
 					resource.TestCheckResourceAttr("bigip_net_vlan.test-vlan", "interfaces.0.vlanport", "1.1"),
@@ -60,11 +60,11 @@ func TestAccBigipNetvlan_import(t *testing.T) {
 		CheckDestroy: testCheckvlansDestroyed,
 		Steps: []resource.TestStep{
 			{
-				Config: TEST_VLAN_RESOURCE,
+				Config: TestVlanResource,
 				Check: resource.ComposeTestCheckFunc(
-					testCheckvlanExists(TEST_VLAN_NAME, true),
+					testCheckvlanExists(TestVlanName, true),
 				),
-				ResourceName:      TEST_VLAN_NAME,
+				ResourceName:      TestVlanName,
 				ImportState:       false,
 				ImportStateVerify: true,
 			},

--- a/docs/resources/bigip_net_vlan.md
+++ b/docs/resources/bigip_net_vlan.md
@@ -39,4 +39,6 @@ resource "bigip_net_vlan" "vlan1" {
 
 * `vlanport` - Physical or virtual port used for traffic
 
+* `cmp_hash` - (Optional,type `string`) Specifies how the traffic on the VLAN will be disaggregated. The value selected determines the traffic disaggregation method. possible options: [`default`, `src-ip`, `dst-ip`]
+
 * `tagged` - Specifies a list of tagged interfaces or trunks associated with this VLAN. Note that you can associate tagged interfaces or trunks with any number of VLANs.

